test/setup/deployers/operator.ts

```diff
@@ -1,19 +1,53 @@
+import * as sleep from 'sleep-promise';
 import { IDeployer, IImageOptions } from './types';
 import * as kubectl from '../../helpers/kubectl';
 
+const PULL_KUBERNETES_OPERATOR_EVENT = 'docker.io/snyk/kubernetes-operator';
 export const operatorDeployer: IDeployer = {
   deploy: deployKubernetesMonitor,
 };
 
+async function seekEvent(event: string): Promise<boolean> {
+  const events = await kubectl.getEvents('marketplace');
+  const foundEvent = events.indexOf(event) > -1;
+
+  return foundEvent;
+}
+
+async function waitToDeployKubernetesOperator(): Promise<void> {
+  while(true) {
+    await kubectl.deleteDeployment('snyk-operator', 'marketplace');
+    await sleep(5000); // give enough time to k8s to apply the previous yaml
+
+    const found = await seekEvent(PULL_KUBERNETES_OPERATOR_EVENT);
+    if (found) {
+      break;
+    }
+  }
+}
+
 async function deployKubernetesMonitor(
   _imageOptions: IImageOptions,
 ): Promise<void> {
-  await kubectl.applyK8sYaml('./test/fixtures/operator/operator-source.yaml');
-  await kubectl.applyK8sYaml('./test/fixtures/operator/installation.yaml');
+  const namespaces = await kubectl.getNamespaces();
+  if (namespaces.includes('openshift-marketplace')) {
+    await kubectl.applyK8sYaml('./test/fixtures/operator/operator-source.yaml');
+    await kubectl.applyK8sYaml('./test/fixtures/operator/installation.yaml');
+    // Await for the Operator to become available, only then
+    // the Operator can start processing the custom resource.
+    await kubectl.waitForDeployment('snyk-operator', 'openshift-marketplace');
+    await kubectl.applyK8sYaml('./test/fixtures/operator/custom-resource.yaml');
+  } else {
+    await kubectl.applyK8sYaml('./test/fixtures/operator/operator-source-k8s.yaml');
+    await kubectl.applyK8sYaml('./test/fixtures/operator/installation-k8s.yaml');
 
-  // Await for the Operator to become available, only then
-  // the Operator can start processing the custom resource.
-  await kubectl.waitForDeployment('snyk-operator', 'snyk-monitor');
+    // Await for the Operator to become available, only then
+    // the Operator can start processing the custom resource.
+    await kubectl.waitForDeployment('snyk-operator', 'marketplace');
+    await kubectl.waitForCRD('snykmonitors.charts.helm.k8s.io');
+    await kubectl.applyK8sYaml('./test/fixtures/operator/custom-resource-k8s.yaml');
+    await waitToDeployKubernetesOperator();
+    await kubectl.waitForDeployment('snyk-operator', 'marketplace');
+  }
```

Comment:
Hmm it almost feels like we could have two distinct deployers - `openshift-deployer.ts` and `olm-deployer.ts`, they seem distinct enough to warrant their own deployment method! This would avoid doing these `if` checks and doing completely different things based on the namespaces we have in the cluster.